lib/reek/configuration/app_configuration.rb

```diff
@@ -5,89 +5,108 @@ module Reek
+      # for anything else.
+      # @return [Hash | nil] the configuration for the source or nil
+      def directory_directive_for_source(source_via)
+        return unless source_via
+        return unless Source::SourceCode.source_from_file?(source_via)
+        base_path = Pathname.new(source_via).split[0]
+        hit = directory_directives.keys.detect { |pathname| pathname.to_s =~ /#{base_path}/ }
+        directory_directives[hit] unless hit.nil?
+      end
 
-          begin
-            @configuration = YAML.load_file(path) || {}
-          rescue => error
-            raise_error(error.to_s, path)
-          end
+      private
 
-          raise_error('Not a hash', path) unless @configuration.is_a? Hash
-        end
+      attr_writer :exclude_paths, :default_directive, :directory_directives
 
-        def reset
-          @configuration.clear
-          @exclude_paths = nil
-        end
+      def load(options)
+        configuration_file = ConfigurationFileFinder.find_and_load(options: options)
 
-        def exclude_paths
-          @exclude_paths ||= begin
-            @configuration.fetch(EXCLUDE_PATHS_KEY, []).map do |string|
-              Pathname.new(string.chomp('/'))
+        configuration_file.each do |key, value|
+          case key
+          when EXCLUDE_PATHS_KEY
+            handle_exclude_paths(value)
+          else
+            klass = try_to_recognize_smell_type key
+            if klass
+              handle_default_directive klass, value
+            else
+              handle_directory_directive key, value
             end
           end
         end
+      end
 
-        private
-
-        def for_smell_types
-          @configuration.reject { |key, _value| NON_SMELL_TYPE_KEYS.include?(key) }
-        end
-
-        def load_smell_type(name)
-          Reek::Smells.const_get(name)
-        rescue NameError
-          report_problem("\"#{name}\" is not a code smell")
-          nil
+      def handle_exclude_paths(paths)
+        self.exclude_paths = paths.map do |path|
+          pathname = Pathname.new path.chomp('/')
+          raise ArgumentError, "Excluded directory #{path} does not exists" unless pathname.exist?
+          pathname
         end
+      end
 
-        def report_problem(reason, path)
-          $stderr.puts "Warning: #{message(reason, path)}"
-        end
+      def handle_default_directive(klass, config)
+        default_directive[klass] = config
+      end
 
-        def raise_error(reason, path)
-          raise ConfigFileException, message(reason, path)
+      def handle_directory_directive(path, config)
+        pathname = Pathname.new path.chomp('/')
+        raise ArgumentError, 'Configuration error:'\
+          "No smell or path for #{pathname}" unless pathname.exist?
+        directory_directives[pathname] = config.each_with_object({}) do |(key, value), hash|
+          hash[Reek::Smells.const_get(key)] = value
```

Comment:
How about "You are trying to configure smell type #{....} but we can't find one with that name. Please make sure you spelled it right. In doubt, check 'config/defaults.reek' in the reek repository for a list of all available smell types."
?